### ISSUE TITLE
Epic: proper vacuuming for databases that are not always on

### ISSUE THREAD

**Issue by kelvich** [neondatabase/neon#6228]
## Motivation

With scale to zero it is easy to end up in the situation where auctovacuum will not run at all. Postgres autovacuum is designed to spread the load over a long period of time to avoid database impact, so for endpoint which are not always on we may never complete vacuum. That will result in a slower performance and bigger startup time (due to clog/multixact bloat).

## DoD

We ensure that one way or another vacuum runs periodically on all of the branches. We should monitor that and define a metric that shows us how severely given database needs vacuum. One example of such metric is `lastTransactionId - min(relfrozenxid)`. Or may be we can pick something better.

## Implementation ideas

The most straightforward approach is to treat autovacuum process the same way as user connections in `compute_ctl` and keep updating `last_active_at` metric while vacuum is running. That would prevent control plane from suspending this endpoint and we'll be sure that vacuum completes. However, due to postgres autovacuum settings we'll keep compute running for unnecessary long time with autovacuum making really slow progress. We could change this setting but then we'll create problems for users run endpoint 24/7.

I can see few options there. Let's imagine `compute_ctl` detected that right now there are no user queries right now and there is a necessity to run vacuum (either by checking presence of active autovacuum or by looking at the same metrics that autovacuum looks at when it decides to start). Now we could:
* change autovacuum settings to the aggressive ones. If user connects while we are vacuuming we should change that setting back 
* start running per-table `vacuum` commands manually. If user connects while we are vacuuming we should stop issuing new vacuum commands and perhaps wait for the currently running vacuums to complete.

First approach looks better since it allows swifter transition between slow-burning vacuum and all-in vacuum. But there are few open questions:
* will running autovacuum pick up new settings? (I assume yes, but needs validation)
* can we configure autovacuum to run on a full speed without any throttling? (I assume yes, but needs validation)

## UI

It would be good to show the user reason for his endpoint to be active right now. We could expose some `time_spend_in_vacuum` field in `compute_ctl` and UI can show that vacuum is in progress near the latest start operation. When vacuum finishes we can show cumulative time that it was running. That should also work in cases when our vacuum was interrupted by user connection and rolled back vacuum setting to the default ones. Makes sence to count only excessive vacuuming time where we were using aggressive settings. Or in other words we show the amount of compute time that we've burnt on vacuum without other user activity (and in theory we can subtract auto suspend time).

### Tasks
- [x] https://github.com/neondatabase/neon/issues/6684
- [x] https://github.com/neondatabase/neon/pull/10995
- [ ] Add metric like 'lastTransactionId - min(relfrozenxid)'?
- [ ] Add an alert (and probably Grafana panel) for cases when we cannot write stats because of the file size limit
- [ ] Rollout to prod region by region (be careful as running autovacuum prevents compute suspend)

## Other related info
- https://www.postgresql.org/docs/16/routine-vacuuming.html

**Comment by jcsp** [comment]
Idea: we could combine this with handling of idle tenants in the rest of the stack, e.g. for an idle tenant pageserver would like to be woken up to GC the tenant after 7 days.  We might want to schedule data integrity scrubs in future too: maybe an overall "scrub" control plane operation could fire up the tenant and then call into all layers of the stack with a general "please do whatever housekeeping you need to do" API hook.

**Comment by vadim2404** [comment]
@knizhnik merged the patch that disabled compute suspension if vacuum works

this week: to investigate the situation with `stats` file

**Comment by kelvich** [comment]
@vadim2404 @knizhnik  did you read what is written in the issue? please revert https://github.com/neondatabase/neon/commit/31a4eb40b2163b3f90439f820af541a7a5e4e970 asap, it will blow up production epically

**Comment by vadim2404** [comment]
I didn't notice any spikes in # of active computes in different regions (we started it with staging) and also @Bodobolero looked at this PR. 

https://neonprod.grafana.net/d/e6971550-2dc3-4e31-aeb0-849ed06841f6/active-computes-by-region?orgId=1&from=now-24h&to=now&viewPanel=1

**Comment by kelvich** [comment]
hm, I missed that it is on prod. Ok, let's keep an eye on th number of active compute instead of rolling back now. But the point still holds -- it may run for dozen of hours, there is no visibility on why it is running, etc.

**Comment by knizhnik** [comment]
Autovacuum is definitely configured to run slowly. But as far as I know there are no some artificial delays (i.e. sleeps...).
It just give up if it fails to set some lock. 

I never saw Postgres where auto vacuum was always active.
Especially in case when there is no database  activity.
Database suspend time is 5 minutes. If autovacuum was not able to complete in 5 minutes for small database (and most of our databases are actually small), then there is something wrong.

**Comment by hlinnaka** [comment]
> But as far as I know there are no some artificial delays (i.e. sleeps...).

There is, see `autovacuum_vacuum_cost_delay` and https://www.postgresql.org/docs/current/runtime-config-resource.html#RUNTIME-CONFIG-RESOURCE-VACUUM-COST.

With a small database, it will still finish reasonably quickly with the default settings though, despite the throttling.

**Comment by vadim2404** [comment]
@knizhnik to look into this issue

**Comment by Bodobolero** [comment]
@kelvich see update in https://github.com/neondatabase/neon/issues/6684

I verified that auto vacuum is no longer interrupted by compute suspend once it is started and thus I verified https://github.com/neondatabase/neon/pull/6322 is working correctly, however without preserving pg_stat cumulative statistics this is only half solved and https://github.com/neondatabase/neon/pull/6613 is important in this context.

**Comment by Bodobolero** [comment]
We also have a problem with monitoring auto-vacuum in Neon https://github.com/neondatabase/neon/issues/6686

**Comment by vadim2404** [comment]
It was partly released, and other items are not so urgent

**Comment by ololobus** [comment]
This week:

- [ ] Waiting on review from @hlinnaka, who is mostly OOO this week, so will likely slip to the next oen

**Comment by ololobus** [comment]
This week:

- [ ] @hlinnaka to have a look, but may slip to the next week again due to traveling

**Comment by ololobus** [comment]
This week:

- [ ] Most of the discussed items are addressed, so @hlinnaka please have another pass
- [ ] @knizhnik @hlinnaka @ololobus to discuss metric async, but log in compute works for the time being

**Comment by ololobus** [comment]
This week:

- [ ] @knizhnik waiting for a minor release, then merge
- [ ] @knizhnik set limit to 16 MB on staging

**Comment by ololobus** [comment]
This week:

- [ ] @ololobus + @knizhnik figure out 1-2 vacuum metrics. Stas proposed '`min(last_autovacuum)` could be an option, but we should avoid any metrics that require connecting to each DB

**Comment by knizhnik** [comment]
Metric `lastTransactionId - min(relfrozenxid)` is IMHOalmost useless. It even worser than "average temperature in hospital" because last one has not sense, but at least some clear semantic (everybody knowns that 36.6 is norm, 39.9 is fever and 18.0 is corpse). Let's say that this metrics returns 1234? Is it bad or normal? Or before stat persisting it was 9999 and now 1111? Or visa versa? It mostly depends on workload than on autovacuum policy.

What else we can measure? There are last_autovacuum and autvacuum_count in pg_stat_all_tables  view.
It seems to be more useful. Especially if considered together with n_ins_since_vacuum
So if logic will be something like this:
```
select now()-min(last_autovacuum) from pg_stat_all_tables join pg_class c on relid=c.oid WHERE 
n_dead_tup > :autovacuum_vacuum_threshold:*:autovacuum_vacuum_scale_factor:*relsize OR
n_ins_since_vacuum > :autovacuum_vacuum_insert_threshold:*autovacuum_vacuum_insert_scale_factor:*relsize;
```

where `autovacuum_vacuum_threshold, autovacuum_vacuum_scale_factor, autovacuum_vacuum_insert_threshold:*, autovacuum_vacuum_insert_scale_factor` are Postgres settings with default values:

```
autovacuum_vacuum_threshold = 50	# min number of row updates before vacuum
autovacuum_vacuum_insert_threshold = 1000	# min number of row inserts before vacuum
autovacuum_vacuum_scale_factor = 0.2	# fraction of table size before vacuum
autovacuum_vacuum_insert_scale_factor = 0.2	# fraction of inserts over table size before insert vacuum
```

then we will get something which may be used to estimate efficiency of persisting statistics.

(note: thew values above can be overwritten for individual tables).

**Comment by knizhnik** [comment]
Concerning concerns that persisting pgstat state can trigger more auto vacuums =>  prevent compute suspension => raise number of active computes, it will be useful to monitor number of tenants which are not suspended because of autovacuum activity. 

The logic of compute suspension is in monitor.cs:
```
                //
                // Do not suspend compute if autovacuum is running
                //
                let autovacuum_count_query = "select count(*) from pg_stat_activity where backend_type = 'autovacuum worker'";
                match cli.query_one(autovacuum_count_query, &[]) {
                    Ok(r) => match r.try_get::<&str, i64>("count") {
                        Ok(num_workers) => {
                            if num_workers > 0 {
                                compute.update_last_active(Some(Utc::now()));
                                continue;
                            }
                        }
```
But I have no idea how top expose it as metric.

**Comment by ololobus** [comment]
This week:

- [ ] @ololobus + @knizhnik figure out 1-2 vacuum metrics. Maybe Peter's recent test is good enough if we run it periodically? https://github.com/neondatabase/neon/pull/10995

**Comment by Bodobolero** [comment]
> Maybe Peter's recent test is good enough if we run it periodically?
It is run periodically every day. See https://github.com/neondatabase/neon/blob/3a0518aee1831f28a3a07d06f052ab0f9d3958b4/.github/workflows/benchmarking.yml#L177

**Comment by ololobus** [comment]
We got stuck discussing the metrics, so plan is to revisit this topic during offsite and decide on the rollout plan

**Comment by hlinnaka** [comment]
This Epic has diverged a lot from the original problem. I'm closing this, opened new issue to track the remaining task of rolling out of the pgstat persistence: https://github.com/neondatabase/cloud/issues/27296.
